File: app/types/query-result.ts
/**
 * クエリ実行結果（Rust側QueryResultと対応）
 */
export interface QueryExecuteResult {
  /** カラム情報 */
  columns: QueryResultColumn[]
  /** 行データ */
  rows: QueryResultRow[]
  /** 取得行数 */
  rowCount: number
  /** 実行時間（ミリ秒） */
  executionTimeMs: number
  /** 警告メッセージ */
  warnings: string[]
}

/**
 * カラム情報
 */
export interface QueryResultColumn {
  /** カラム名 */
  name: string
  /** データ型 */
  dataType: string
  /** NULL許容 */
  nullable: boolean
}

/**
 * 行データ
 */
export interface QueryResultRow {
  /** 値の配列（カラム順） */
  values: QueryValue[]
}

/**
 * 値の型
 */
export type QueryValue = null | boolean | number | string | number[]

/**
 * クエリ実行リクエスト
 */
export interface QueryExecuteRequest {
  /** 接続ID */
  connectionId: string
  /** 実行するSQL */
  sql: string
  /** タイムアウト（秒） */
  timeoutSeconds?: number
}

/**
 * クエリ実行レスポンス
 */
export interface QueryExecuteResponse {
  /** クエリID（キャンセル用） */
  queryId: string
  /** 実行結果 */
  result: QueryExecuteResult
}

/**
 * クエリエラー
 */
export interface QueryExecuteError {
  /** エラーコード */
  code: QueryErrorCode
  /** エラーメッセージ */
  message: string
  /** 詳細情報 */
  details?: QueryErrorDetails
}

export type QueryErrorCode =
  | 'connection_failed'
  | 'query_timeout'
  | 'query_cancelled'
  | 'syntax_error'
  | 'permission_denied'
  | 'table_not_found'
  | 'column_not_found'
  | 'unknown'

export interface QueryErrorDetails {
  /** エラー位置（行） */
  line?: number
  /** エラー位置（列） */
  column?: number
  /** エラー発生箇所のSQL */
  sqlSnippet?: string
}

File: src-tauri/src/models/query_result.rs
use serde::{Deserialize, Serialize};

/// クエリ実行結果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryResult {
    /// カラム情報
    pub columns: Vec<QueryResultColumn>,
    /// 行データ
    pub rows: Vec<QueryResultRow>,
    /// 取得行数
    pub row_count: usize,
    /// 実行時間（ミリ秒）
    pub execution_time_ms: u64,
    /// 警告メッセージ（ある場合）
    pub warnings: Vec<String>,
}

/// カラム情報
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryResultColumn {
    /// カラム名
    pub name: String,
    /// データ型
    pub data_type: String,
    /// NULL許容
    pub nullable: bool,
}

/// 行データ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryResultRow {
    /// 値の配列（カラム順）
    pub values: Vec<QueryValue>,
}

/// 値の型
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum QueryValue {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    String(String),
    Bytes(Vec<u8>),
    // 日付・時刻は文字列でシリアライズ
}

/// クエリ実行リクエスト
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryExecuteRequest {
    /// 接続ID
    pub connection_id: String,
    /// 実行するSQL
    pub sql: String,
    /// タイムアウト（秒、オプション）
    pub timeout_seconds: Option<u32>,
}

/// クエリ実行レスポンス
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryExecuteResponse {
    /// クエリID（キャンセルトークン用）
    pub query_id: String,
    /// 実行結果
    pub result: QueryResult,
}

/// クエリエラー
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryError {
    /// エラーコード
    pub code: QueryErrorCode,
    /// エラーメッセージ
    pub message: String,
    /// 詳細情報（SQL位置など）
    pub details: Option<QueryErrorDetails>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum QueryErrorCode {
    ConnectionFailed,
    QueryTimeout,
    QueryCancelled,
    SyntaxError,
    PermissionDenied,
    TableNotFound,
    ColumnNotFound,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryErrorDetails {
    /// エラー位置（行）
    pub line: Option<u32>,
    /// エラー位置（列）
    pub column: Option<u32>,
    /// エラー発生箇所のSQL
    pub sql_snippet: Option<String>,
}
